querytest/compiler.go

```diff
@@ -9,52 +9,51 @@ import (
 	"github.com/influxdata/flux/stdlib/influxdata/influxdb/v1"
 )
 
-// FromCSVCompiler wraps a compiler and replaces all From operations with FromCSV
-type FromCSVCompiler struct {
+// ReplaceSpecFunction is a function that takes an Operation and changes its spec
+type ReplaceSpecFunction func(op *flux.Operation)
+
+type ReplaceSpecCompiler struct {
 	flux.Compiler
-	InputFile string
+	fn ReplaceSpecFunction
 }
 
-// FromInfluxJSONCompiler wraps a compiler and replaces all From operations with FromJSON
-type FromInfluxJSONCompiler struct {
-	flux.Compiler
-	InputFile string
+func NewReplaceCompiler(compiler flux.Compiler, fn ReplaceSpecFunction) *ReplaceSpecCompiler {
+	return &ReplaceSpecCompiler{Compiler: compiler, fn: fn}
 }
 
-func (c FromCSVCompiler) Compile(ctx context.Context) (*flux.Spec, error) {
+func (c *ReplaceSpecCompiler) Compile(ctx context.Context) (*flux.Spec, error) {
 	spec, err := c.Compiler.Compile(ctx)
 	if err != nil {
 		return nil, err
 	}
-	ReplaceFromSpec(spec, c.InputFile)
+	c.replaceSpec(spec)
 	return spec, nil
 }
 
-func (c FromInfluxJSONCompiler) Compile(ctx context.Context) (*flux.Spec, error) {
-	spec, err := c.Compiler.Compile(ctx)
-	if err != nil {
-		return nil, err
+func (c *ReplaceSpecCompiler) replaceSpec(q *flux.Spec) {
+	for _, op := range q.Operations {
+		c.fn(op)
 	}
-	ReplaceFromWithFromInfluxJSONSpec(spec, c.InputFile)
-	return spec, nil
 }
 
-func ReplaceFromSpec(q *flux.Spec, csvSrc string) {
-	for _, op := range q.Operations {
+// FromCSVCompiler wraps a compiler and replaces all From operations with FromCSV
+func FromCSVCompiler(c flux.Compiler, csvFile string) *ReplaceSpecCompiler {
+	return NewReplaceCompiler(c, func(op *flux.Operation) {
 		if op.Spec.Kind() == influxdb.FromKind {
 			op.Spec = &csv.FromCSVOpSpec{
-				File: csvSrc,
+				File: csvFile,
 			}
 		}
-	}
+	})
 }
 
-func ReplaceFromWithFromInfluxJSONSpec(q *flux.Spec, jsonSrc string) {
-	for _, op := range q.Operations {
+// FromInfluxJSONCompiler wraps a compiler and replaces all From operations with FromJSON
+func FromInfluxJSONCompiler(c flux.Compiler, jsonFile string) *ReplaceSpecCompiler {
+	return NewReplaceCompiler(c, func(op *flux.Operation) {
 		if op.Spec.Kind() == influxdb.FromKind {
 			op.Spec = &v1.FromInfluxJSONOpSpec{
-				File: jsonSrc,
+				File: jsonFile,
 			}
 		}
-	}
+	})
```

Comment:
This change is required in order to let InfluxDB change the operation spec according to its kinds and not Flux's.

Reply:
An interesting thing about this piece of code is that actually it is unused in Flux! Should I move it towards InfluxDB?

Reply:
👍 Ok for this version? Or the previous one?

Reply:
Not sure what you mean?  We capture 1.x data as JSON, and `fromInfluxJSON` can read it and produce Flux tables.

Reply:
I wasn't clear, sorry. I mean, should I move this version to InfluxDB or the previous one? This one is more general in that it lets specify generic `ReplaceCompiler`s.

Reply:
Oh, got it. 😄   I like the refactoring you did for `ReplaceSpecCompiler` and it seems like it belongs in `flux`.  I think you can just move `FromInfluxJSONCompiler` to `influxdb`.